apps/www/_blog/2023-08-11-supavisor-1-million.mdx

```diff
@@ -0,0 +1,436 @@
+  />
+</div>
+
+In the Supavisor architecture, only single node holds direct connections to each database instance. If you add Postgres read-replicas, or another databases, then the Supavisor cluster will spread the connections to the replicas. Every Supavisor instance can accept incoming connections and either execute queries themselves (if they directly connected) or relay to another node (if not).
+
+This setup successfully handled:
+
+- **1,003,200 simultaneous connections** to the Supavisor instances.
+- **20,000 QPS** or **1.2 million queries per minute.** Each user connected executed `select` queries once every 50 seconds.
+
+<div>
+  <img
+    alt="chart connections 1 million light"
+    className="dark:hidden"
+    src="/images/blog/2023-08-11-supavisor-1-million/chart-1m-light.png"
+  />
+  <img
+    alt="chart connections 1 million dark"
+    className="hidden dark:block"
+    src="/images/blog/2023-08-11-supavisor-1-million/chart-1m-dark.png"
+  />
+</div>
+
+Within the cluster:
+
+- The directly connected instance was under almost the same load as when handling 500,000 concurrent clients in a single-node mode.
+- The relaying instance was extremely over-resourced. Most cores had little-to-no workload because relayed connections are more lightweight.
+
+<div>
+  <img
+    alt="cpu load 1 million light"
+    className="dark:hidden"
+    src="/images/blog/2023-08-11-supavisor-1-million/cpu-1m-light.png"
+  />
+  <img
+    alt="cpu load 1 million dark"
+    className="hidden dark:block"
+    src="/images/blog/2023-08-11-supavisor-1-million/cpu-1m-dark.png"
+  />
+</div>
+
+In a multitenant setup (or when using Read-Replicas), the load is much more evenly spread because all Supavisor instances connect to comparable numbers of databases and have both direct and relayed connections evenly distributed between each other.
+
+<div>
+  <img
+    alt="diagram supavisor 2 instances with more databases light"
+    className="dark:hidden"
+    src="/images/blog/2023-08-11-supavisor-1-million/diagram-2db-light.png"
+  />
+  <img
+    alt="diagram supavisor 2 instances with more databases dark"
+    className="hidden dark:block"
+    src="/images/blog/2023-08-11-supavisor-1-million/diagram-2db-dark.png"
+  />
+</div>
+
+### Supavisor's impact on query duration
+
+To measure the impact, we started with 5,000 QPS (queries per second). This allows us to exclude side effects from the database side (long query execution times).
+
+The query used in the experiment was the following:
+
+```sql
+select *
+from (
+    values
+    (1, 'one'),
+    (2, 'two'),
+    (3, 'three')
+) as t (num, letter);
+```
+
+We found with Supavisor median query duration was less than 2ms. And this includes not only time from client to Supavisor but the whole roundtrip: from Client to Supavisor ➡️ from Supavisor to Postgres ➡️ then query execution time on Postgres ➡️ and back to Supavisor ➡️ and to the Client.
```

Comment:
I just don't know how this is possible?

> from Client to Supavisor

Even this would take longer than 2ms? Was the execution happening on the same server?

Reply:
Same aws region

Reply:
We tried ping from supavisor to db instance and it shows submiliisecond timings, just networking should be around 1ms for the whole roundtrip

Reply:
Maybe "client" is confusing because people associate "client" with "browser"?

Reply:
ok to me, client is just client, do you have smth on your mind that would fit better?